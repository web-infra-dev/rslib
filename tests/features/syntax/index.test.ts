import { join } from 'node:path';
import { buildAndGetResults } from 'test-helper';
import { expect, test } from 'vitest';

test('should downgrade class private method by default', async () => {
  const fixturePath = join(__dirname, 'default');
  const { entries } = await buildAndGetResults(fixturePath);

  expect(entries.esm).toMatchSnapshot();
  expect(entries.esm).toContain('#bar');

  expect(entries.cjs).toContain('#bar');
});

test('should downgrade class private method with output.syntax config', async () => {
  const fixturePath = join(__dirname, 'config');
  const { entries } = await buildAndGetResults(fixturePath);

  expect(entries.esm0).toMatchSnapshot();
  expect(entries.esm0).not.toContain('#bar');

  expect(entries.esm1).toMatchSnapshot();
  expect(entries.esm1).toContain('#bar');

  expect(entries.cjs0).toContain('#bar');
  expect(entries.cjs0).toContain('()=>{'); // test webpack runtime arrow function
  expect(entries.cjs1).not.toContain('#bar');
  expect(entries.cjs1).not.toContain('()=>{'); // test webpack runtime arrow function
});
