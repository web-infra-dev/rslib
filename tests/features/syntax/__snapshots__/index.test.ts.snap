// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should downgrade class private method by default 1`] = `
"class Foo {
    constructor(){
        this.#bar();
    }
    #bar() {}
}
export { Foo };
"
`;

exports[`should downgrade class private method with output.syntax config 1`] = `
"function _check_private_redeclaration(obj, privateCollection) {
    if (privateCollection.has(obj)) throw new TypeError(\\"Cannot initialize the same private elements twice on an object\\");
}
function _class_private_method_get(receiver, privateSet, fn) {
    if (!privateSet.has(receiver)) throw new TypeError(\\"attempted to get private field on non-instance\\");
    return fn;
}
function _class_private_method_init(obj, privateSet) {
    _check_private_redeclaration(obj, privateSet);
    privateSet.add(obj);
}
var _bar = /*#__PURE__*/ new WeakSet();
class Foo {
    constructor(){
        _class_private_method_init(this, _bar);
        _class_private_method_get(this, _bar, bar).call(this);
    }
}
function bar() {}
export { Foo };
"
`;

exports[`should downgrade class private method with output.syntax config 2`] = `
"class Foo {
    constructor(){
        this.#bar();
    }
    #bar() {}
}
export { Foo };
"
`;
